src/encoder/stateless/h264.rs

```diff
@@ -295,86 +244,18 @@ where
 
         Ok(())
     }
-
-    fn poll_pending(&mut self, mode: BlockingMode) -> EncodeResult<()> {
-        // Poll the output queue once and then continue polling while new promise is submitted
-        while let Some(coded) = self.output_queue.poll(mode)? {
-            self.coded_queue.push_back(coded);
-        }
-
-        while let Some(recon) = self.recon_queue.poll(mode)? {
-            let requests = self.predictor.reconstructed(recon)?;
-            if requests.is_empty() {
-                // No promise was submitted, therefore break
-                break;
-            }
-
-            for request in requests {
-                self.execute(request)?;
-            }
-        }
-
-        Ok(())
-    }
 }
 
-impl<H, B> StatelessVideoEncoder<H> for StatelessEncoder<H, B>
+impl<Handle, Backend> StatelessEncoder<H264, Handle, Backend>
 where
-    B: StatelessH264EncoderBackend,
-    B: StatelessEncoderBackendImport<H, B::Picture>,
+    Backend: StatelessH264EncoderBackend,
+    Backend: StatelessEncoderBackendImport<Handle, Backend::Picture>,
 {
-    fn encode(&mut self, metadata: FrameMetadata, handle: H) -> EncodeResult<()> {
-        log::trace!(
-            "encode: timestamp={} layout={:?}",
-            metadata.timestamp,
-            metadata.layout
-        );
-
-        // Import `handle` to backends representation
-        let backend_pic = self.backend.import_picture(&metadata, handle)?;
-
-        // Increase the number of frames that predictor holds, before handing one to it
-        self.predictor_frame_count += 1;
-
-        // Ask predictor to decide on the next move and execute it
-        let requests = self.predictor.new_frame(backend_pic, metadata)?;
-        for request in requests {
-            self.execute(request)?;
-        }
-
-        Ok(())
-    }
-
-    fn drain(&mut self) -> EncodeResult<()> {
-        log::trace!("currently predictor holds {}", self.predictor_frame_count);
-
-        // Drain the predictor
-        while self.predictor_frame_count > 0 || !self.recon_queue.is_empty() {
-            if self.output_queue.is_empty() && self.recon_queue.is_empty() {
-                // The OutputQueue is empty and predictor holds frames, force it to yield a request
-                // to empty it's internal queue.
-                let requests = self.predictor.drain()?;
-                self.predictor_frame_count -= requests.len();
-
-                for request in requests {
-                    self.execute(request)?;
-                }
-            }
-
-            self.poll_pending(BlockingMode::Blocking)?;
-        }
-
-        // There are still some requests being processed. Continue on polling them.
-        while !self.output_queue.is_empty() {
-            self.poll_pending(BlockingMode::Blocking)?;
-        }
-
-        Ok(())
-    }
+    fn new_h264(backend: Backend, config: EncoderConfig, mode: BlockingMode) -> EncodeResult<Self> {
```

Comment:
I guess we could replace these codec-specific methods with a single one if we add a few more associated types (`EncoderConfig` and `Predictor`) to `StatelessCodec`, but let's do this as a follow-up.